play-json/shared/src/main/scala-3/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -78,7 +76,7 @@ object JsMacroImpl {
   inline def readCasesL[A: Reads, L <: Tuple, T <: Tuple](js: JsValue, name: String): JsResult[A] =
     inline (erasedValue[L], erasedValue[T]) match {
       case _: (l *: ls, t *: ts)       =>
-        if (name == typeName[l]) summonReads[t & A].reads(js)
+        if (name == typeName[l, A]) summonReads[t & A].reads(js)
```

Comment:
Unrelated with this PR but why need to use `t & A`?

Reply:
> cant we use `t` instead of `A`?

Uhm, cos I'm an idiot apparently.  I'll try that 😄

> also I think we can remove the `A: Reads` constraint

That constraint is necessary for recursive class definitions, for which there are examples in the test.

> Unrelated with this PR but why need to use `t & A`?

`t & A` is so we can avoid doing `summonReads[t].asInstanceOf[Reads[A]]` which is required in order to have a `JsResult[A]` result type.  So we summon the intersection.  Ideally I'd put this in the definition of `t` but I don't know how to define `t *: ts` and `t <: A` in one go... 🤔 I can do it in two nested inline matches, using an intermediate type variable name, but that doesn't seem that much better than the intersection.

Reply:
> That constraint is necessary for recursive class definitions, for which there are examples in the tests.

I tried removing them in master branch and tests still pass :thinking: . Going to send a PR to see if travis agrees...

Reply:
Make sure to remove existing classfiles, because the incremental compiler is optimistic and doesn't recompile inline defs... 

```bash
rm play-json/jvm/target/scala-3.0.0-RC1/test-classes/play/api/libs/json/MacroSpec*
```

Reply:
well travis likes it https://github.com/playframework/play-json/pull/573